spec/operations/save_operation_spec.cr

```diff
@@ -157,6 +161,54 @@ describe "Avram::SaveOperation" do
     operation.changes[:name].should be_nil
   end
 
+  describe "upsert upsert_lookup_columns" do
+    describe ".upsert!" do
+      it "updates the existing record if one exists" do
+        existing_user = UserFactory.create &.name("Rich").nickname(nil).age(20)
+        joined_at = Time.utc.at_beginning_of_second
+
+        user = UpsertUserOperation.upsert!(
+          name: "Rich",
+          nickname: nil,
+          age: 30,
+          joined_at: joined_at
+        )
+
+        UserQuery.new.select_count.should eq(1)
+        user = user.not_nil!
+        user.id.should eq(existing_user.id)
+        user.name.should eq("Rich")
+        user.nickname.should be_nil
+        user.age.should eq(30)
+        user.joined_at.should eq(joined_at)
+      end
+
+      it "creates a new record if one doesn't exist" do
+        existing_user = UserFactory.create &.name("Rich").nickname(nil).age(20)
+        joined_at = Time.utc.at_beginning_of_second
+
+        user = UpsertUserOperation.upsert!(
+          name: "Rich",
+          nickname: "R.",
+          age: 30,
+          joined_at: joined_at
+        )
```

Comment:
So this says "Look up a user by name = "Rich" and nickname = "R." which doesn't exist, right? With the bang method, should this raise an exception like `Avram::RecordNotFound`? Or maybe `user` should be `nil` here?

Reply:
With the bang version it should only raise if the record is invalid. Otherwise it will create a new one if the record is not found. I should add a test for that though!